src/observers/stream_observer.ts

```diff
@@ -58,19 +60,19 @@ export class StreamObserver {
 
   receiveMessageEvent = (event: MessageEvent) => {
     if (this.started && typeof event.data == "string") {
-      this.receiveMessageHTML(event.data)
+      this.receiveMessageHTML(event.data, "event")
```

Comment:
As an alternative to making the Streams aware of their origin, how do you feel about adding or extending the `Session` behavior for dispatching `Request` instances or receiving `Response` instances?

Something like a `BusyStateObserver` that listens for `turbo:before-fetch-request` and `turbo:before-fetch-response` events. The `Session` could be the delegate for one instance, and our existing code for Form and Frame `[aria-busy]` toggling could be abstracted into their own instances.

Reply:
Yeah, I like that idea. I'll give that a shot and see how it looks 👍

Reply:
@kevinmcconnell after suggesting that and sitting with the idea for a while, I've realized that the current behavior might be the most correct.

When a Frame is navigating, it's marked as `[aria-busy="true"]`. When a `<form>` is submitting, it's marked as `[aria-busy="true"]`. When the entire page is redirected, the `<body>` is marked as `[aria-busy="true"]`.

For a Turbo Stream request that's targeted at a small fragment of the document, nesting the entire page within an `[aria-busy="true"]` element would pose accessibility issues, since it'd communicate to assistive technologies that the full document is in a state of flux, which isn't the case.

With that in mind, I think we're better off with the current behavior. Am I misunderstanding your intent?

Reply:
Ok, I'm sorry to misunderstand _again_. I think a more correct resolution to this issue would be to _prevent_ the document from ever setting `[aria-busy="true"]` in the first place, instead of finding a way to properly clear that attribute.

Does that sound like a reasonable alternative?

Reply:
We can certainly avoid setting the `aria-busy` attribute in this case, if that's preferred. I actually wondered about that initially, but I didn't go that route for a couple of reasons:

1. If you have a link with `data-turbo-stream`, it's still valid to get an HTML response. In which case the behaviour is a regular visit, and so it might be expected that the busy state behaviour matches that of a regular visit.
2. It can still be useful to show a busy state even if you're updating part of the page (as we do with frame, forms, etc). When the link is outside of a frame or form it's harder to know what the scope of that change is going to be, so setting the attribute on the document element seems like a reasonable thing to do.

I agree with your point that the attribute is communicating that the document is in a state of flux, but without having a defined boundary of the part that's changing, then I'm not sure where better to indicate this.

If we want to say that `data-turbo-stream` doesn't set busy state because of this, I think that's OK, but it might leave people without such an easy way to show a loading spinner or similar during that request.

Reply:
> When the link is outside of a frame or form it's harder to know what the scope of that change is going to be, so setting the attribute on the document element seems like a reasonable thing to do.

In the context of full page navigations, marking the entire document as busy is reasonable, since all content is likely to be swapped out.

From the specification for [aria-busy](https://www.w3.org/TR/wai-aria/#aria-busy):

> Indicates an element is being modified and that **assistive technologies MAY want to wait until the modifications are complete before exposing them to the user**.

From the section on [Live Regions](https://www.w3.org/TR/wai-aria/#dfn-live-region):

> Since these asynchronous areas are expected to update outside the user's area of focus, assistive technologies such as screen readers have either been unaware of their existence or unable to process them for the user. WAI-ARIA has provided a collection of properties that allow the author to identify these live regions and process them: aria-live, aria-relevant, aria-atomic, and aria-busy.

From the section on [Required Owned Elements]()

> When a widget is missing required owned elements due to script execution or loading, authors MUST mark a containing element with [aria-busy](https://www.w3.org/TR/wai-aria/#aria-busy) equal to true. For example, **until a page is fully initialized and complete, an author could mark the document element as busy**.

Marking the entire document as busy would make its descendant content inaccessible to assistive tech, which has the potential to be very jarring unless a Turbo Stream response targets the entire `<body>` element.

> If you have a link with `data-turbo-stream`, it's still valid to get an HTML response. In which case the behaviour is a regular visit, and so it might be expected that the busy state behaviour matches that of a regular visit.

This is certainly an issue.

When considering `<form>` submissions that are Stream-enabled, the operation is scoped to the `<form>` element itself, which sets `[aria-busy=]` throughout the lifecycle of the request. It doesn't face the same uncertainty, since a submission response is either `2xx` or `4xx` (operates on the current page) or a `3xx` redirect, which is _for certain_ a `Visit`. The `Visit` lifecycle kicks off, and its activity bubbles up throughout the document, and is reflected in the `html[aria-busy]` attribute.

Given the nature of `<a data-turbo-stream>` elements, that ambiguity is baked-in.

I've historically been strongly opposed to `<a data-turbo-stream>`, so that bias might be obscuring me from a solution for that ambiguity. With that being said, I am hopeful that there is a solution, and am happy to help implement if we're able to discover it.

Reply:
Yes, I agree it would be good to find a solution. Appreciate your feedback on this. (And I also appreciate we may have different opinions on whether `<a data-turbo-stream>` is a good idea ;))

One idea could be to allow marking a boundary where a busy state should apply (and default to not setting `aria-busy` if no such boundary is present). That way people could define behaviour that's appropriate for their use case, without much effort.

I just started looking at `aria-live` to see if that would be a valid way to approach this. Something like: when requesting a stream update from within a Live Region we'd mark that region as busy. Otherwise we'd fall back to current behaviour, only setting it on the `form`, `frame` etc where appropriate). But I'm not yet very familiar with `aria-live` so I don't know if that's actually a misguided approach.

Reply:
I think the nature of `<turbo-steam>` responses makes any kind of "geographic" limitation on the client-side moot, since a response can target any part of the document, regardless of where the request was initiated from.

Reply:
I think that's just an issue with stream responses in general (not with `data-turbo-stream` specifically). Responses to form submissions can target areas outside the form (and might only do so); actions that are initiated from inside a `turbo-frame` can similarly update areas outside of that frame, etc. Anything can reach outside its own scope, and since `aria-busy` is communicating what's being modified, it seems like the only "safe" place to set it would be the document element. But as you point out, it may not always be desirable to do that.

I think that's why it could be useful to have a simple way for people to override the default behaviour, since the default behaviour won't always be right.

Another consideration is that `aria-busy` is really about what's being modified, but often what people want to show in their UI (and what they use the attribute for) is which operation is pending. e.g. if you have a button that causes another section of the UI to change, you may want to show some sort of spinner or pending state on the button, rather than on that part of the UI that will be mutated.